lib/internal/modules/helpers.js

```diff
@@ -320,6 +329,28 @@ function normalizeReferrerURL(referrerName) {
 }
 
 
+let esmSyntaxErrorMessagesSet; // Declared lazily in shouldRetryAsESM
+let throwsOnlyInCommonJSErrorMessagesSet; // Declared lazily in shouldRetryAsESM
+/**
+ * After an attempt to parse a module as CommonJS throws an error, should we try again as ESM?
```

Comment:
How do we make this informed decision? It might help to add a single line explanation to here